monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/EvalOnStartOperator.scala

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2014-2018 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.reactive.internal.operators
+
+import monix.eval.Task
+import monix.execution.Ack
+import monix.execution.Ack.Stop
+import monix.reactive.Observable.Operator
+import monix.reactive.observers.Subscriber
+
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+
+private[reactive] final
+class EvalOnStartOperator[A](cb: A => Task[Unit])
+  extends Operator[A,A] {
+
+  def apply(out: Subscriber[A]): Subscriber[A] =
+    new Subscriber[A] {
+      implicit val scheduler = out.scheduler
+
+      private[this] var isDone = false
+      private[this] var isStart = true
+
+      def onNext(elem: A): Future[Ack] = {
+        // Protects calls to user code from within the operator and
+        // stream the error downstream if it happens, but if the
+        // error happens because of calls to `onNext` or other
+        // protocol calls, then the behavior should be undefined.
+        try {
+          if (isStart) {
+            val t = for {
+              _ <- cb(elem).onErrorHandle(ex => { onError(ex); Stop })
```

Comment:
I'm not sure about this. @alexandru please correct me if I'm wrong.

Here we call `out.OnError` in case of error, return `Stop` but ignore it and still call `out.OnNext` which violates `Observer` contract (you shouldn't send anything after `onError`)

In my opinion we should short citcuit here in case of error. Ideally, we could write failing test now and then make it work after fix. Something similar to what you already have with hand written `Subscriber` but keeping track of `onNext` / `onError` calls

Reply:
Thanks @Avasil I've noticed this bug and was thinking about how to fix it after I pushed yesterday :D
And short-circuting in case of error makes perfect sense, I just couldn't figure out how to implement it in a best way.